cmake/macros/SundialsAddBenchmark.cmake

```
@@ -0,0 +1,88 @@
+# Take in an executable, determine how to construct the arguments
+
+# ---------------------------------------------------------------
+# CMake macro for adding benchmarks to `make benchmark`.
+# ---------------------------------------------------------------
+
+macro(sundials_add_benchmark NAME EXECUTABLE BASE_BENCHMARK_NAME)
+
+  # Define single value parameters the macro takes in to set up the test runner
+  #
+  # NUM_CORES         = number of cores (GPU count or CPU count) to run on/number of resource sets
+  # BENCHMARK_ARGS    = arguments to pass to the executable
+  # IDENTIFIER        = suffix to append to end of benchmark name
+  set(oneValueArgs NUM_CORES BENCHMARK_ARGS IDENTIFIER)
+
+  # TEST_RUNNER_ARGS  = command line arguments to pass to the test executable
+  set(multiValueArgs TEST_RUNNER_ARGS )
+
+  # ENABLE_GPU        = indicate this benchmark should be run with GPUs
+  set(options ENABLE_GPU)
+
+  cmake_parse_arguments(sundials_add_benchmark
+    "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  # set the target name
+  if(sundials_add_benchmark_IDENTIFIER)
+    set(TARGET_NAME ${NAME}_${sundials_add_benchmark_IDENTIFIER})
+  else()
+    if(sundials_add_benchmark_BENCHMARK_ARGS)
+      string(REPLACE " " "_" TEST_SUFFIX "${sundials_add_benchmark_BENCHMARK_ARGS}")
+      set(TARGET_NAME ${NAME}_${TEST_SUFFIX})
+    else()
+      set(TARGET_NAME ${NAME}_run)
+    endif()
+  endif()
+
+  # Create default benchmark caliper output directory if custom directory is not defined
+  if(SUNDIALS_CALIPER_OUTPUT_DIR)
+    set(SUNDIALS_BENCHMARK_OUTPUT_DIR ${SUNDIALS_CALIPER_OUTPUT_DIR}/Benchmarking/${BASE_BENCHMARK_NAME})
+  else()
+    set(SUNDIALS_BENCHMARK_OUTPUT_DIR ${PROJECT_BINARY_DIR}/Benchmarking/${BASE_BENCHMARK_NAME})
+  endif()
+
+  # make the output directory if it doesn't exist
+  if(NOT EXISTS ${SUNDIALS_BENCHMARK_OUTPUT_DIR}/${TARGET_NAME})
+    file(MAKE_DIRECTORY ${SUNDIALS_BENCHMARK_OUTPUT_DIR}/${TARGET_NAME})
+  endif()
+
+  # command line arguments for the test runner script
+  set(TEST_RUNNER_ARGS
+    "--verbose"
+    "--executablename=$<TARGET_FILE:${EXECUTABLE}>"
+    "--outputdir=${TEST_OUTPUT_DIR}"
+    "--calidir=${SUNDIALS_BENCHMARK_OUTPUT_DIR}/${TARGET_NAME}"
+    "--nodiff"
+    )
+  
+  # incorporate scheduler arguments into test_runner
+  if(SUNDIALS_SCHEDULER_COMMAND STREQUAL "flux run")
+    set(SCHEDULER_STRING " -n${sundials_add_benchmark_NUM_CORES}")
+  elseif(SUNDIALS_SCHEDULER_COMMAND STREQUAL "jsrun" AND ${sundials_add_benchmark_ENABLE_GPU})
+    set(SCHEDULER_STRING " --smpiargs=\\\"-gpu\\\" -n${sundials_add_benchmark_NUM_CORES} -a1 -c1 -g1")
```

```suggestion
    set(SCHEDULER_STRING " --smpiargs=\\\"-gpu\\\" -n${sundials_add_benchmark_NUM_GPUS} -a1 -c1 -g1")
```

Since we usually run with number of tasks equal to number of GPUs when using GPUs, lets do this instead for this case.

---

For this one, it's NUM_CORES as that number varys depending on if you pass in the GPU variable or the CPU variable. Is there a case where the benchmarks will be run with both CPUs and GPUs? I noticed from the `build_and_bench` script it was either one or the other.